lfs/diff_index_scanner.go

```diff
@@ -0,0 +1,191 @@
+	// the index.
+	DstName string
+}
+
+// DiffIndexScanner scans the output of the `git diff-index` command.
+type DiffIndexScanner struct {
+	// next is the next entry scanned by the Scanner.
+	next *DiffIndexEntry
+	// err is any error that the Scanner encountered while scanning.
+	err error
+
+	// from is the underlying scanner, scanning the `git diff-index`
+	// command's stdout.
+	from *bufio.Scanner
+}
+
+// NewDiffIndexScanner initializes a new `DiffIndexScanner` scanning at the
+// given ref, "ref".
+//
+// If "cache" is given, the DiffIndexScanner will scan for differences between
+// the given ref and the index. If "cache" is _not_ given, DiffIndexScanner will
+// scan for differences between the given ref and the currently checked out
+// tree.
+//
+// If any error was encountered in starting the command or closing its `stdin`,
+// that error will be returned immediately. Otherwise, a `*DiffIndexScanner`
+// will be returned with a `nil` error.
+func NewDiffIndexScanner(ref string, cached bool) (*DiffIndexScanner, *wrappedCmd, error) {
+	cmd, err := startCommand("git", diffIndexCmdArgs(ref, cached)...)
+	if err != nil {
+		return nil, cmd, errors.Wrap(err, "diff-index")
+	}
+
+	if err = cmd.Stdin.Close(); err != nil {
+		return nil, cmd, errors.Wrap(err, "diff-index: close")
+	}
+
+	return &DiffIndexScanner{
+		from: bufio.NewScanner(cmd.Stdout),
+	}, cmd, nil
+}
+
+// diffIndexCmdArgs returns a string slice containing the arguments necessary
+// to run the diff-index command.
+func diffIndexCmdArgs(ref string, cached bool) []string {
+	args := []string{"diff-index", "-M"}
+	if cached {
+		args = append(args, "--cached")
+	}
+	args = append(args, ref)
+
+	return args
+}
+
+// Scan advances the scan line and yields either a new value for Entry(), or an
+// Err(). It returns true or false, whether or not it can continue scanning for
+// more entries.
+func (s *DiffIndexScanner) Scan() bool {
+	if !s.prepareScan() {
+		return false
+	}
+
+	s.next, s.err = s.scan(s.from.Text())
+
+	return s.err == nil
+}
+
+// Entry returns the last entry that was Scan()'d by the DiffIndexScanner.
+func (s *DiffIndexScanner) Entry() *DiffIndexEntry { return s.next }
+
+// Entry returns the last error that was encountered by the DiffIndexScanner.
+func (s *DiffIndexScanner) Err() error { return s.err }
+
+// prepareScan clears out the results from the last Scan() loop, and advances
+// the internal scanner to fetch a new line of Text().
+func (s *DiffIndexScanner) prepareScan() bool {
+	s.next, s.err = nil, nil
+	if !s.from.Scan() {
+		s.err = s.from.Err()
+		return false
+	}
+
+	return true
+}
+
+// scan parses the given line and returns a `*DiffIndexEntry` or an error,
+// depending on whether or not the parse was successful.
+func (s *DiffIndexScanner) scan(line string) (*DiffIndexEntry, error) {
+	parts := strings.Split(line, "\t")
+	if len(parts) < 2 {
+		return nil, errors.Errorf("invalid line: %s", line)
```

Comment:
These errors are also generic, and could potentially be extracted into types. I think this is a bit overkill for the `lfs` package (think: `lfs.DiffIndexScannerInvalidLineErr` 😅 ), but I think would be appropriate for a new package entirely.

That being said, their messages are too generic, and should at least be wrapped with `diff-index: `. I think the most appropriate place to do this would be in the public `Scan()` function from which this function is called.